Detectors/Vertexing/src/SVertexer.cxx

```diff
@@ -436,29 +448,49 @@ bool SVertexer::checkV0(const TrackCand& seedP, const TrackCand& seedN, int iP,
 }
 
 //__________________________________________________________________
-int SVertexer::checkCascades(float rv0, std::array<float, 3> pV0, float p2V0, int avoidTrackID, int posneg, int ithread)
+int SVertexer::checkCascades(float rv0, std::array<float, 3> pV0, float p2V0, int avoidTrackID, int posneg, VBracket v0vlist, int ithread)
 {
+
   // check last added V0 for belonging to cascade
   auto& fitterCasc = mFitterCasc[ithread];
   const auto& v0 = mV0sTmp[ithread].back();
   auto& tracks = mTracksPool[posneg];
-  const auto& pv = mPVertices[v0.getVertexID()];
   int nCascIni = mCascadesTmp[ithread].size();
-  // start from the 1st track compatible with V0's primary vertex
-  int firstTr = mVtxFirstTrack[posneg][v0.getVertexID()], nTr = tracks.size();
-  if (firstTr < 0) {
-    firstTr = nTr;
+
+  // check if a given PV has already been used in a cascade
+  std::unordered_map<int, int> pvMap;
+
+  // start from the 1st bachelor track compatible with V0's primary vertex
+  // minimum intersection of V0 and bachelor track --> min V0 vert and max bachelor vert
+  auto v0Idx = v0vlist.getMin();
+  int firstIdx = mVtxMaxLUT[posneg][v0Idx];
+
+  while (firstIdx < 0 || firstIdx >= mTrackSortVtxMax[posneg].size()) {
+    v0Idx++;
+    firstIdx = mVtxMaxLUT[posneg][v0Idx];
   }
+
+  int firstTr = mTrackSortVtxMax[posneg][firstIdx], nTr = tracks.size();
+
   for (int it = firstTr; it < nTr; it++) {
-    if (it == avoidTrackID) {
+    auto& trackInd = mTrackSortVtxMax[posneg][it];
```

Comment:
@fmazzasc The `firstTr` is first extracted from the mTrackSortVtxMax, then you pass it again to `mTrackSortVtxMax`, why? Should not you simply loop like?
```
for (int it = firstIdx; it < nTr; it++)
{
  int firstTr = mTrackSortVtxMax[posneg][it];
}
```

Reply:
You are right, thanks! I also spotted a bug in the LUT table, that I have fixed. You can see the impact of the fix in the attached plot
![c1](https://user-images.githubusercontent.com/43742195/181486244-e431b16d-6328-40f3-bf7b-b36d5f4e558b.png)